Sources/JSONUtilities/Dictionary+JSONKey.swift

```diff
@@ -174,159 +126,172 @@ extension Dictionary where Key: StringProtocol {
+  }
 
+  fileprivate func JSONArrayForKey(atKeyPath keyPath: Key) throws -> JSONArray {
+    return try getValue(atKeyPath: keyPath)
   }
 
-  // MARK: JSONDictionary and JSONArray creation
+  // MARK: Value decoding
 
-  fileprivate func JSONDictionaryForKey(atKeyPath keyPath: Key) throws -> JSONDictionary {
-    guard let jsonDictionary = self[keyPath: keyPath] as? JSONDictionary else {
-      throw DecodingError.mandatoryKeyNotFound(key: keyPath)
+  fileprivate func getValue<A, B>(array: [A], value: A) throws -> B {
+    guard let typedValue = value as? B else {
+      throw DecodingError.incorrectTypeInArray(array: array, expectedType: B.self, value: value)
     }
-    return jsonDictionary
+    return typedValue
   }
 
-  fileprivate func JSONArrayForKey(atKeyPath keyPath: Key) throws -> JSONArray {
-    guard let jsonArray = self[keyPath: keyPath] as? JSONArray else {
-      throw DecodingError.mandatoryKeyNotFound(key: keyPath)
+  fileprivate func getValue<T>(atKeyPath key: Key) throws -> T {
+    guard let value = self[key] else {
+      // swiftlint:disable:next force_cast
+      throw DecodingError.keyNotFound(dictionary: self as! JSONDictionary, key: key)
     }
-    return jsonArray
+    guard let typedValue = value as? T else {
+      // swiftlint:disable:next force_cast
+      throw DecodingError.incorrectTypeInDictionary(dictionary: self as! JSONDictionary, key: key, expectedType: T.self, value: value)
+    }
+    return typedValue
   }
 
-  // MARK: JSONArray decoding
+  // MARK: Dictionary decoding
+
+  fileprivate func decodeDictionary<T>(atKeyPath keyPath: Key, invalidItemBehaviour: InvalidItemBehaviour = .remove, decode: (JSONDictionary, String) throws -> T) throws -> [String: T] {
+    let jsonDictionary: JSONDictionary = try json(atKeyPath: keyPath)
+
+    var dictionary: [String: T] = [:]
+    for (key, _) in jsonDictionary {
 
-  fileprivate func decodableObjectsArray<ReturnType: JSONObjectConvertible>(_ jsonArray: JSONArray) -> [ReturnType] {
-    return jsonArray.flatMap {
-      guard let castedJsonObject = $0 as? JSONDictionary else {
-        return nil
+      switch invalidItemBehaviour {
+      case .remove:
+        if let typedItem = try? decode(jsonDictionary, key) {
+          dictionary[key] = typedItem
+        }
+      case .fail:
+        dictionary[key] = try decode(jsonDictionary, key)
       }
+    }
+
+    return dictionary
+  }
 
-      return try? ReturnType(jsonDictionary: castedJsonObject)
+  // MARK: Array decoding
+
+  fileprivate func decodeArray<T>(atKeyPath keyPath: Key, invalidItemBehaviour: InvalidItemBehaviour = .remove, decode: (JSONArray, Any) throws -> T) throws -> [T] {
+    let jsonArray = try JSONArrayForKey(atKeyPath: keyPath)
+
+    return try jsonArray.flatMap {
+      switch invalidItemBehaviour {
+      case .remove:
+        guard let typedItem  = try? decode(jsonArray, $0) else {
+          return nil
+        }
+        return typedItem
```

Comment:
Yes, sorry that was from before when decode returned an optional, so doing try? returned a double optional. Now everything just throws as we have more error information